### PR TITLE
Added unit tests for the Notification Model

### PR DESCRIPTION
**Pull request by ChrisCRL**
### Added Unit Testing for Notification model
---

### **What does this PR do?**  

This PR introduces a robust unit tests for our Django-based notification system. The tests validate the Notification model, REST API endpoints, utility functions, and WebSocket integration using Django Channels. 

---

### **What has been added or changed?**  
✔ Testing: Comprehensive test suite with mock objects for WebSocket and API endpoint testing.

---

### **What problem does this solve?**  

Unit testing for notifications solves several important problems in software development:

- **Reliability verification:** Tests ensure that notifications are delivered correctly to the intended recipients.
- **Integration validation:** Testing WebSocket functionality verifies that real-time notification delivery works properly.
- **Regression prevention:** As the codebase evolves, tests catch unintended changes that might break notification functionality.
- **Edge case handling:** Tests verify that the system handles unusual situations properly, like scheduled notifications or marking multiple notifications as read simultaneously.
- **API contract enforcement:** Tests ensure that the notification API endpoints maintain their expected behavior and response formats.
- **Mock dependency testing:** The tests demonstrate that the notification system works properly even when external dependencies (like the WebSocket channel layer) are replaced with mocks.

---

## **Related Issue**  
🔗 **This PR addresses** Issue #67  

---

### **How was this feature developed?**  
- **Django TestCase framework:** foundation for all test cases
- **APIClient from Django REST Framework:** testing API endpoints
- **MagicMock from unittest.mock:** simulating WebSocket channels and async functions

---

## **Design Documentation (If Applicable)**  
**Diagrams & Visuals**
N/A

---

### **Step-by-step testing guide**  
1. Head to backend directory
2. Enable virtual environment.
3. Run `pip install -r requirements.txt`
4. Run `python manage.py migrate`
3. Run `python manage.py test apps.notifications`
---

## **Files Added/Modified**  
📂 **List of important files changed in this PR**  

| **File Name**      | **Description** |
|--------------------|----------------|
| `tests.py`        | Unit Testing for Notifications |
| `requirements.txt`        | Project Libraries |

---

## **AI Code Usage**  
🤖 **Was AI-generated code used in this PR?**  
- [x] Yes - AI was used to aid the initial structure of `tests.py`. 
- [] No (_All code was manually implemented._)  

---

## **Checklist for Reviewers**  
**Things to check before approving this PR:**  
- [] Do all tests pass?

---

### **What improvements can be made after merging?**  
- **Add Unit Testing For Scheduled Notifications**: Currently not implemented. Will be implemented as I work on #50 

---

### **Final Notes**  
**Thank you for reviewing!** Please leave feedback if you have any suggestions or concerns. Let’s make this feature even better!

### CODE REVIEW

File: backend/apps/notifications/tests.py [hunk 1]
```diff
@@ -1,3 +1,262 @@
...
+            'notification_title': 'API Test',
+            'notification_message': 'Testing API creation',
+            'notification_type': 'success',
+            'info_category': 'general'
+        }
+
+        response = self.client.post(url, data)
+
+        print(f"Response status: {response.status_code}")
+        print(f"Response content: {response.content.decode()}")
```
Comment (omgdory) [inline backend/apps/notifications/tests.py:156]: These print lines are cluttering the testing output: should be removed!

**Comment by omgdory** [comment]
In your PR comment, please correct the typo `app.notifications` to `apps.notifications` to avoid confusion. Apart from this and the aforementioned print statements, all tests pass and your tests make sense. The `tasks.py` test, as you mentioned, can be done in a separate PR which addresses Issue #50. In any case, please resolve these two small things and I'll approve!

File: backend/apps/notifications/tests.py (hunk 1 above)
Comment (ChrisCRL) [reply backend/apps/notifications/tests.py:156]: oopsie :P sorry, debugging lines. Thanks for pointing it out!

**Comment by ChrisCRL** [comment]
> In your PR comment, please correct the typo `app.notifications` to `apps.notifications` to avoid confusion. 
ty for pointing this out. It is now changed.

**Review by omgdory — APPROVED** [review]
## ✅ Approval Overview
Testing makes sense, is well-documented, and passes. Requirements satisfied and do not seem to conflict. Well done!

## 🎯 Next Steps
The remaining test associated with the task in tasks.py can be implemented in a later PR. If manual testing/usage proves that the functionality is intact, then we can worry about it later down the line.
